OpenSim/Examples/DataTable/example.cpp

```diff
@@ -0,0 +1,357 @@
+    auto elem2_real = dt_real.getElt(9, "col-four"); ignore(elem2_real);
+    auto elem2_vec3 = dt_vec3.getElt(9, "col-four"); ignore(elem2_vec3);
+    auto elem2_vec6 = dt_vec6.getElt(9, "col-four"); ignore(elem2_vec6);
+  }
+  // Update an element by retrieving it using its row-col pair with upd<>
+  // functions. The upd<> functions return a writable reference to the element.
+  {
+    dt_real.updElt(5, 7) *= 2;
+    dt_vec3.updElt(5, 7) *= 2;
+    dt_vec6.updElt(5, 7) *= 2;
+
+    // Column labels can be used in place of column index, if the columns have
+    // been given labels.
+    dt_real.updElt(9, "col-four") *= 2;
+    dt_vec3.updElt(9, "col-four") *= 2;
+    dt_vec6.updElt(9, "col-four") *= 2;
+  }
+
+  // Get a copy of the underlying matrix.
+  auto mat_real = dt_real.getAsMatrix(); ignore(mat_real);
+  auto mat_vec3 = dt_vec3.getAsMatrix(); ignore(mat_real);
+  auto mat_vec6 = dt_vec6.getAsMatrix(); ignore(mat_real);
+
+  // It is possible to clear the data in the DataTable if we want to fill it up
+  // with other data later. This also clears the column labels.
+  dt_real.clearData();
+  dt_vec3.clearData();
+  dt_vec6.clearData();
+
+  // DataTable can hold metadata in form of key-value pairs where key is always
+  // a std::string and value can be of any type. All DataTables, regardless of
+  // the underlying type, can hold metadata.
+  
+  // Create some objects that represent metadata.
+  int integer{200};
+  std::string string{"metadata"};
+
+  // Insert an int as metadata.
+  dt_real.insertMetaData("integer", integer);
+
+  // Insert a std::string as metadata.
+  dt_real.insertMetaData("string", string);
+
+  // Insert a SimTK::Matrix_ as metadata.
+  dt_real.insertMetaData("simtk_matrix", *mat_real);
+
+  // Insert a std::vector<SimTK::Real> as metadata.
+  dt_real.insertMetaData("data_real", data_real);
+
+  // Insert another DataTable as metadata.
+  dt_real.insertMetaData("datatable", dt_vec3);
+
+
+  // To retrieve the objects stored as metadata, two things have to be provided.
+  // The key under which the object was stored and the exact type of the 
+  // object.
+
+  // Retrieve int from metadata.
+  assert(dt_real.getMetaData<int>("integer") == integer);
+
+  // Retrieve std::string from metadata.
+  assert(dt_real.getMetaData<std::string>("string") == string);
+
+  // Retrieve SimTK::Matrix_ from metadata.
+  const auto& mat_real_copy = 
+    dt_real.getMetaData<SimTK::Matrix_<double>>("simtk_matrix");
+  for(int r = 0; r < mat_real->nrow(); ++r)
+    for(int c = 0; c < mat_real->ncol(); ++c) {
+      auto a = mat_real_copy.getElt(r, c);
+      auto b = mat_real->getElt(r, c);
+      assert(std::abs(a - b) < EPSILON || (std::isnan(a) && std::isnan(b)));
+    }
+
+  // Retrieve std::vector from metadata.
+  auto data_real_copy = dt_real.getMetaData<decltype(data_real)>("data_real");
+
+  // Retrieve the datatable from metadata.
+  auto dt_vec3_cpy = dt_real.getMetaData<decltype(dt_vec3)>("datatable");
```

Comment:
I think that for an example, the use of `decltype` here makes the code more difficult to understand.